Mathlib/Data/Fin/Basic.lean

```diff
@@ -104,231 +212,2385 @@ lemma Fin.ofNat'_succ : {n : Nat} → [Nonempty (Fin n)] →
+
+@[simp]
+theorem lt_sub_one_iff {n : ℕ} {k : Fin (n + 2)} : k < k - 1 ↔ k = 0 := by
+  rcases k with ⟨_ | k, hk⟩
+  simp [lt_iff_val_lt_val]
+  have : (k + 1 + (n + 1)) % (n + 2) = k % (n + 2) := by
+    rw [add_right_comm, add_assoc, add_mod_right]
+  simp [lt_iff_val_lt_val, ext_iff, Fin.coe_sub, succ_eq_add_one, this,
+    mod_eq_of_lt ((lt_succ_self _).trans hk)]
+#align fin.lt_sub_one_iff Fin.lt_sub_one_iff
+
+@[simp]
+theorem le_sub_one_iff {n : ℕ} {k : Fin (n + 1)} : k ≤ k - 1 ↔ k = 0 := by
+  cases n
+  · simp
+  rw [← lt_sub_one_iff, le_iff_lt_or_eq, lt_sub_one_iff, or_iff_left_iff_imp, eq_comm,
+    sub_eq_iff_eq_add]
+  simp
+#align fin.le_sub_one_iff Fin.le_sub_one_iff
+
+@[simp]
+theorem sub_one_lt_iff {n : ℕ} {k : Fin (n + 1)} : k - 1 < k ↔ 0 < k :=
+  not_iff_not.1 <| by simp only [not_lt, le_sub_one_iff, le_zero_iff]
+#align fin.sub_one_lt_iff Fin.sub_one_lt_iff
+
+theorem last_sub (i : Fin (n + 1)) : last n - i = Fin.rev i :=
+  ext <| by rw [coe_sub_iff_le.2 i.le_last, val_last, val_rev, Nat.succ_sub_succ_eq_sub]
+#align fin.last_sub Fin.last_sub
+
+end AddGroup
+
+section SuccAbove
+
+theorem succ_above_aux (p : Fin (n + 1)) :
+    StrictMono fun i : Fin n => if (castSucc i) < p then (castSucc i) else i.succ :=
+  (castSucc : Fin n ↪o _).strictMono.ite (succEmbedding n).strictMono
+    (fun _ _ hij hj => lt_trans ((castSucc : Fin n ↪o _).lt_iff_lt.2 hij) hj) fun i =>
+    (castSucc_lt_succ i).le
+#align fin.succ_above_aux Fin.succ_above_aux
+
+/-- `succAbove p i` embeds `fin n` into `fin (n + 1)` with a hole around `p`. -/
+def succAbove (p : Fin (n + 1)) : Fin n ↪o Fin (n + 1) :=
+  OrderEmbedding.ofStrictMono _ p.succ_above_aux
+#align fin.succ_above Fin.succAbove
+
+/-- Embedding `i : fin n` into `fin (n + 1)` with a hole around `p : fin (n + 1)`
+embeds `i` by `cast_succ` when the resulting `i.cast_succ < p`. -/
+theorem succAbove_below (p : Fin (n + 1)) (i : Fin n) (h : castSucc i < p) :
+    p.succAbove i = castSucc i := if_pos h
+#align fin.succ_above_below Fin.succAbove_below
+
+@[simp]
+theorem succAbove_ne_zero_zero {a : Fin (n + 2)} (ha : a ≠ 0) : a.succAbove 0 = 0 := by
+  rw [Fin.succAbove_below]
+  · simp
+  · simp only [castSucc_zero]
+    exact bot_lt_iff_ne_bot.mpr ha
+#align fin.succ_above_ne_zero_zero Fin.succAbove_ne_zero_zero
+
+theorem succAbove_eq_zero_iff {a : Fin (n + 2)} {b : Fin (n + 1)} (ha : a ≠ 0) :
+    a.succAbove b = 0 ↔ b = 0 := by
+  simp only [← succAbove_ne_zero_zero ha, OrderEmbedding.eq_iff_eq, iff_self]
+#align fin.succ_above_eq_zero_iff Fin.succAbove_eq_zero_iff
+
+theorem succAbove_ne_zero {a : Fin (n + 2)} {b : Fin (n + 1)} (ha : a ≠ 0) (hb : b ≠ 0) :
+    a.succAbove b ≠ 0 :=
+  mt (succAbove_eq_zero_iff ha).mp hb
+#align fin.succ_above_ne_zero Fin.succAbove_ne_zero
+
+/-- Embedding `fin n` into `fin (n + 1)` with a hole around zero embeds by `succ`. -/
+@[simp]
+theorem succAbove_zero : ⇑(succAbove (0 : Fin (n + 1))) = Fin.succ :=
+  rfl
```

Comment:
This works:
```lean
theorem succAbove_zero : ⇑(succAbove (0 : Fin (n + 1))) = Fin.succ := by
  simp [succAbove]
```

Reply:
Answer (according to the porting meeting): we do understand (the definition of `0 : Fin (n + 1)` changed) and it's not clear whether we should press on or try something else.